### PR TITLE
feat: allow message limits (#82)

### LINKED ISSUE #82: Allow remote configuration
vscode-redhat-telemetry allows remote configuration: https://github.com/redhat-developer/vscode-redhat-telemetry?tab=readme-ov-file#remote-configuration

We should allow the same.

### PR DESCRIPTION
**Pull request by adietish**
fixes #82

### CODE REVIEW

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    private final OkHttpClient client = new OkHttpClient.Builder()
+            .connectTimeout(5, TimeUnit.SECONDS)
+            .readTimeout(5, TimeUnit.SECONDS)
+            .writeTimeout(5, TimeUnit.SECONDS)
+            .build();
+
+    public String getRemote() {
+        Request request = new Request.Builder()
+                .url(REMOTE)
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java:46]: I strongly suggest `System.getProperty(REDHAT_TELEMETRY_REMOTE_CONFIG_URL, REMOTE)`, that makes it way easier to test different versions of remote urls

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    static final String EMBEDDED = "/telemetry-config.json";
+    static final String REMOTE = "https://raw.githubusercontent.com/adietish/intellij-redhat-telemetry/issue-82/src/main/resources/telemetry-config.json";
+
+    private final OkHttpClient client = new OkHttpClient.Builder()
+            .connectTimeout(5, TimeUnit.SECONDS)
+            .readTimeout(5, TimeUnit.SECONDS)
+            .writeTimeout(5, TimeUnit.SECONDS)
+            .build();
+
+    public String getRemote() {
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java:44]: maybe check if it's a file:// URL and try to open it directly

Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java:44]: maybe name it fetchRemote(), add some javadoc. Add `@Nullable` annotation to return type

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        return toString(Configurations.class.getResourceAsStream(EMBEDDED));
+    }
+
+    private String toString(InputStream in) {
+        if (in == null) {
+            return null;
+        }
+        BufferedReader reader = new BufferedReader(new InputStreamReader(in));
+        return reader.lines().collect(Collectors.joining());
+    }
```
Comment (fbricon) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java:93]: use try-with-resource

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+import java.nio.file.attribute.FileTime;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+class Configurations {
+
+    private static final Logger LOGGER = Logger.getInstance(Configurations.class);
+    static final Path LOCAL = Directories.RED_HAT.resolve("telemetry-config.json");
+    static final String EMBEDDED = "/telemetry-config.json";
+    static final String REMOTE = "https://raw.githubusercontent.com/adietish/intellij-redhat-telemetry/issue-82/src/main/resources/telemetry-config.json";
```
[the remark this answers is not in the document]
Comment (adietish) [reply src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java:36]: @fbricon there's no config in main yet. It'll only exist once this PR is merged.

**Comment by adietish** [comment]
@fbricon, @sbouchet: to test this PR in a plugin you can use  https://github.com/redhat-developer/intellij-kubernetes/pull/765

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (fbricon) [reply src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/Configurations.java:36]: that's why it's important to be able to override the url with a system property

**Comment by adietish** [comment]
@fbricon corrected what you suggested. I also fixed proper fallback to embedded config if remote/local are not available. Please re-review.

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/ClasspathConfiguration.java [hunk 5]
```diff
@@ -10,6 +10,8 @@
  ******************************************************************************/
 package com.redhat.devtools.intellij.telemetry.core.configuration;
 
+import org.jetbrains.annotations.Nullable;
```
Comment (sbouchet) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/ClasspathConfiguration.java:13]: seems this import is not used.

File: src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/LimitsConfigurations.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+import static com.redhat.devtools.intellij.telemetry.core.util.FileUtils.ensureExists;
+import static com.redhat.devtools.intellij.telemetry.core.util.FileUtils.getPathForFileUrl;
+
+class LimitsConfigurations {
+
+    private static final Logger LOGGER = Logger.getInstance(LimitsConfigurations.class);
+    static final Path LOCAL = Directories.RED_HAT.resolve("telemetry-config.json");
+    static final String EMBEDDED = "/telemetry-config.json";
+    static final String REMOTE = "https://raw.githubusercontent.com/adietish/intellij-redhat-telemetry/issue-82/src/main/resources/telemetry-config.json";
```
Comment (sbouchet) [inline src/main/java/com/redhat/devtools/intellij/telemetry/core/configuration/limits/LimitsConfigurations.java:42]: Seems this line should change to points to 'main'

**Comment by fbricon** [comment]
Tested it with lsp4ij and it failed:

```
2024-07-04 12:00:35,870 [ 371591] SEVERE - #c.i.i.p.PluginManager - 'void com.redhat.devtools.intellij.telemetry.core.service.TelemetryMessageBuilder.<init>(java.lang.ClassLoader)'
java.lang.NoSuchMethodError: 'void com.redhat.devtools.intellij.telemetry.core.service.TelemetryMessageBuilder.<init>(java.lang.ClassLoader)'
	at com.redhat.devtools.lsp4ij.internal.telemetry.RedHatTelemetryService.lambda$new$0(RedHatTelemetryService.java:32)
	at com.redhat.devtools.intellij.telemetry.core.util.Lazy.get(Lazy.java:27)
	at com.redhat.devtools.lsp4ij.internal.telemetry.RedHatTelemetryService.getMessageBuilder(RedHatTelemetryService.java:45)
	at com.redhat.devtools.lsp4ij.internal.telemetry.RedHatTelemetryService.send(RedHatTelemetryService.java:35)
	at com.redhat.devtools.lsp4ij.internal.telemetry.TelemetryManager$1.send(TelemetryManager.java:108)
	at com.redhat.devtools.lsp4ij.internal.telemetry.TelemetryManager$1.lambda$handleAdded$0(TelemetryManager.java:79)
	at java.base/java.util.Collections$SingletonSet.forEach(Collections.java:4905)
	at com.redhat.devtools.lsp4ij.internal.telemetry.TelemetryManager$1.handleAdded(TelemetryManager.java:79)
```

**Comment by fbricon** [comment]
Can we avoid breaking API/ABI ?

**Comment by adietish** [comment]
@fbricon:

> Can we avoid breaking API/ABI ?

This API breakage is caused by the fix for issue #824
We could keep the existing API but would keep plugins report the wrong exception. IMHO it is preferrable that plugins fix thing on their side.

**Comment by adietish** [comment]
> Can we avoid breaking API/ABI ?

Re-added the old constructor which then works with lsp4ij. Testing the other plugins that we have.

**Comment by adietish** [comment]
I tested the binary version of this plugin successfully with the following plugins:
* kubernetes
* quarkus
* lsp4ij
* openshift
* rsp
* dependency-analytics
* tekton
* knative

IMHO compatibility is no issue any more.

**Review by sbouchet — APPROVED** [review]
tested locally with openshift extension. LGTM

**Review by fbricon — APPROVED** [review]

**Pull request merged**